sphinx/theming.py

```diff
@@ -11,212 +11,143 @@
-        return
 
-    def __init__(self, name):
-        # type: (unicode) -> None
-        if name not in self.themes:
-            self.load_extra_theme(name)
-            if name not in self.themes:
-                if name == 'sphinx_rtd_theme':
-                    raise ThemeError('sphinx_rtd_theme is no longer a hard dependency '
-                                     'since version 1.4.0. Please install it manually.'
-                                     '(pip install sphinx_rtd_theme)')
-                else:
-                    raise ThemeError('no theme named %r found '
-                                     '(missing theme.conf?)' % name)
+class Theme(object):
+    def __init__(self, name, theme_path, factory):
+        # type: (unicode, unicode, HTMLThemeFactory) -> None
         self.name = name
+        self.base = None
+        self.rootdir = None
 
-        # Do not warn yet -- to be compatible with old Sphinxes, people *have*
-        # to use "default".
-        # if name == 'default' and warn:
-        #     warn("'default' html theme has been renamed to 'classic'. "
-        #          "Please change your html_theme setting either to "
-        #          "the new 'alabaster' default theme, or to 'classic' "
-        #          "to keep using the old default.")
-
-        tdir, tinfo = self.themes[name]
-        if tinfo is None:
+        if path.isdir(theme_path):
             # already a directory, do nothing
             self.rootdir = None
-            self.themedir = tdir
-            self.themedir_created = False
+            self.themedir = theme_path
         else:
             # extract the theme to a temp directory
             self.rootdir = tempfile.mkdtemp('sxt')
             self.themedir = path.join(self.rootdir, name)
-            self.themedir_created = True
-            ensuredir(self.themedir)
-            for name in tinfo.namelist():
-                if name.endswith('/'):
-                    continue
-                dirname = path.dirname(name)
-                if not path.isdir(path.join(self.themedir, dirname)):
-                    os.makedirs(path.join(self.themedir, dirname))
-                with open(path.join(self.themedir, name), 'wb') as fp:
-                    fp.write(tinfo.read(name))
-
-        self.themeconf = configparser.RawConfigParser()
-        self.themeconf.read(path.join(self.themedir, THEMECONF))  # type: ignore
+            extract_zip(theme_path, self.themedir)
+
+        self.config = configparser.RawConfigParser()
+        self.config.read(path.join(self.themedir, THEMECONF))  # type: ignore
 
         try:
-            inherit = self.themeconf.get('theme', 'inherit')
+            inherit = self.config.get('theme', 'inherit')
         except configparser.NoOptionError:
-            raise ThemeError('theme %r doesn\'t have "inherit" setting' % name)
+            raise ThemeError(_('theme %r doesn\'t have "inherit" setting') % name)
```

Comment:
I think that ERROR messages should be output in the console's language setting.

Sphinx's `_(...)` determines the language according to the conf.py language setting. Therefore, when Klingon language document is being built, error messages will be displayed in Klingon language whatever the console language is. (I can't read it)